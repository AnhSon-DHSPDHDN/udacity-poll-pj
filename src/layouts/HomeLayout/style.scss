.home-layout {
  display: flex;
  justify-content: center;

  &__main-page {
    margin-top: 60px;
    padding-top: 30px;
    width: 100%;
    max-width: 1200px;
  }
}
