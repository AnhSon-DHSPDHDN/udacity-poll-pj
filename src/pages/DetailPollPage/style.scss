@import '../../styles/variables';

.detail-poll-page {
  display: flex;
  flex-direction: column;
  align-items: center;

  &__avatar {
    width: 250px;
    height: 250px;
    border-radius: 100%;
  }

  &__questions-container {
    display: flex;
    column-gap: 20px;
  }

  .question {
    display: flex;
    width: 400px;
    border: 1px solid $color-black;
    flex-direction: column;

    &__name {
      padding: 7px 20px;
    }

    &__btn-choose {
      outline: none;
      font-weight: bold;
      background-color: green;
      border: none;
      padding: 10px 0;
      cursor: pointer;
      transition: .3s;

      &:hover {
        background-color: rgb(88, 139, 11);
      }

      &:active {
        background-color: rgb(49, 80, 3);
      }

      &:disabled {
        cursor: not-allowed;
        background-color: rgb(88, 139, 11);
      }
    }

    &--not-choose {
      opacity: .2;
    }
  }
}
