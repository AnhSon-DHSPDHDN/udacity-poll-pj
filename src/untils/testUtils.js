import { render } from "@testing-library/react";
import { Provider } from "react-redux";
import { MemoryRouter } from "react-router-dom";
import store from "../redux/store";

export function renderWithAllProvider(
  ui,
  { initialEntries, ...renderOptions }
) {
  function Wrapper({ children }) {
    return (
      <MemoryRouter initialEntries={initialEntries}>
        <Provider store={store}>{children}</Provider>
      </MemoryRouter>
    );
  }

  return render(ui, { wrapper: Wrapper, ...renderOptions });
}
