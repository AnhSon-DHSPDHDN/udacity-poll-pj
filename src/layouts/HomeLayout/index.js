import React from "react";
import { useDispatch, useSelector } from "react-redux";
import { Navigate, Outlet, useLocation } from "react-router-dom";
import HomeHeader from "../../components/HomeHeader";
import Routes from "../../constants/routes";
import { actUpdateCallbackUrl } from "../../redux/features/auth/authSlice";

import "./style.scss";

const HomeLayout = () => {
  const { isAuth } = useSelector((state) => state.auth);
  const location = useLocation();
  const dispatch = useDispatch();

  if (!isAuth) {
    dispatch(actUpdateCallbackUrl(location.pathname));
    return <Navigate to={Routes.LOGIN_PAGE} />;
  }

  return (
    <div className="home-layout">
      <HomeHeader />
      <div className="home-layout__main-page">
        <Outlet />
      </div>
    </div>
  );
};

export default HomeLayout;
