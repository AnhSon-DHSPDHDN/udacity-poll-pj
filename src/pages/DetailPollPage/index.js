import React, { useEffect, useMemo } from "react";
import { useDispatch, useSelector } from "react-redux";
import { useNavigate, useParams } from "react-router-dom";
import Routes from "../../constants/routes";
import {
  actFetchAnswerQuestion,
  actFetchQuestions,
} from "../../redux/features/questions/questionsSlice";
import "./style.scss";

const Question = ({ option, name, question, currentUser, isDisableAnswer }) => {
  const dispatch = useDispatch();

  const handleAnswerQuestion = (questionName) => {
    const payload = {
      authedUser: currentUser.id,
      qid: question.id,
      answer: questionName,
    };

    dispatch(actFetchAnswerQuestion(payload));
  };

  return (
    <div className="question">
      <span className="question__name">{option?.text}</span>
      <button
        className="question__btn-choose"
        onClick={() => handleAnswerQuestion(name)}
        disabled={isDisableAnswer}
      >
        Click
      </button>
    </div>
  );
};

const DetailPollPage = () => {
  const { question_id } = useParams();
  const dispatch = useDispatch();
  const navigate = useNavigate();
  const { questions, isLoading, isLoadingAnswer } = useSelector(
    (state) => state.questions
  );
  const { users, userInfo } = useSelector((state) => state.auth);

  useEffect(() => {
    dispatch(actFetchQuestions());
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);

  useEffect(() => {
    if (Object.keys(questions).length && !questions[question_id]) {
      navigate(Routes.NOT_FOUND);
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [questions, question_id]);

  const currentQuestion = useMemo(
    () => questions[question_id] || {},
    [question_id, questions]
  );

  const isHasAnswer = [
    ...(currentQuestion?.optionOne?.votes || []),
    ...(currentQuestion?.optionTwo?.votes || []),
  ].includes(userInfo.id);

  const isDisableAnswer = isLoadingAnswer || isLoading || isHasAnswer;

  const computedRateChooseOption = (question, name) => {
    const totalAnswerNumber =
      [...question?.optionOne?.votes, ...question?.optionTwo?.votes].length ||
      1;
    const currentAnswerNumber = question?.[name]?.votes.length || 0;

    return (currentAnswerNumber * 100) / totalAnswerNumber + " %";
  };

  return (
    <div className="detail-poll-page">
      <span className="detail-poll-page__create-by text-bold">
        Poll by {currentQuestion.author}
      </span>
      <img
        className="detail-poll-page__avatar mt-4 mb-4"
        alt="avatar"
        src={users[currentQuestion.author]?.avatarURL}
      />
      <span className="detail-poll-page__title mb-4 text-bold">
        Would You Rather?
      </span>
      <div className="detail-poll-page__questions-container">
        <div>
          <Question
            option={currentQuestion.optionOne}
            question={currentQuestion}
            name={"optionOne"}
            currentUser={userInfo}
            isDisableAnswer={isDisableAnswer}
          />
          {isHasAnswer && (
            <>
              <p className="text-error">
                Number of people who chose the answer:{" "}
                {currentQuestion.optionOne.votes?.length || 0}
              </p>
              <p className="text-error">
                Answer selection rate:{" "}
                {computedRateChooseOption(currentQuestion, "optionOne")}
              </p>
            </>
          )}
        </div>
        <div>
          <Question
            option={currentQuestion.optionTwo}
            question={currentQuestion}
            name={"optionTwo"}
            currentUser={userInfo}
            isDisableAnswer={isDisableAnswer}
          />
          {isHasAnswer && (
            <>
              <p className="text-error">
                Number of people who chose the answer:{" "}
                {currentQuestion.optionTwo.votes?.length || 0}
              </p>
              <p className="text-error">
                Answer selection rate:{" "}
                {computedRateChooseOption(currentQuestion, "optionTwo")}
              </p>
            </>
          )}
        </div>
      </div>
    </div>
  );
};

export default DetailPollPage;
