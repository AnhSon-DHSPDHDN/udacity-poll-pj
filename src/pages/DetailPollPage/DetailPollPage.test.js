import { screen } from "@testing-library/react";
import { Route, Routes } from "react-router-dom";
import DetailPollPage from ".";
import { renderWithAllProvider } from "../../untils/testUtils";
import * as redux from "react-redux";
import HomeLayout from "../../layouts/HomeLayout";
import { act } from "react-dom/test-utils";

jest.mock("react-redux", () => {
  return {
    __esModule: true,
    ...jest.requireActual("react-redux"),
  };
});

describe("Detail Poll Page", () => {
  beforeEach(() => {
    const mockStateRedux = {
      questions: {
        questions: {
          am8ehyc8byjqgar0jgpub9: {
            id: "am8ehyc8byjqgar0jgpub9",
            author: "sarahedo",
            timestamp: 1488579767190,
            optionOne: {
              votes: [],
              text: "be telekinetic",
            },
            optionTwo: {
              votes: ["sarahedo"],
              text: "be telepathic",
            },
          },
        },
        isLoading: false,
        isLoadingAnswer: false,
      },
      auth: {
        users: {
          sarahedo: {
            id: "sarahedo",
            password: "123456",
            name: "Sarah Edo",
            avatarURL: `${process.env.PUBLIC_URL}/imgs/avt_1.jpeg`,
            answers: {
              "8xf0y6ziyjabvozdd253nd": "optionOne",
              "6ni6ok3ym7mf1p33lnez": "optionOne",
              am8ehyc8byjqgar0jgpub9: "optionTwo",
              loxhs1bqm25b708cmbf3g: "optionTwo",
            },
            questions: ["8xf0y6ziyjabvozdd253nd", "am8ehyc8byjqgar0jgpub9"],
          },
          tylermcginnis: {
            id: "tylermcginnis",
            password: "123456",
            name: "Tyler McGinnis",
            avatarURL: `${process.env.PUBLIC_URL}/imgs/avt_3.jpeg`,
            answers: {
              vthrdm985a262al8qx3do: "optionOne",
              xj352vofupe1dqz9emx13r: "optionTwo",
            },
            questions: ["loxhs1bqm25b708cmbf3g", "vthrdm985a262al8qx3do"],
          },
          johndoe: {
            id: "johndoe",
            password: "123456",
            name: "John Doe",
            avatarURL: `${process.env.PUBLIC_URL}/imgs/avt_2.jpeg`,
            answers: {
              xj352vofupe1dqz9emx13r: "optionOne",
              vthrdm985a262al8qx3do: "optionTwo",
              "6ni6ok3ym7mf1p33lnez": "optionOne",
            },
            questions: ["6ni6ok3ym7mf1p33lnez", "xj352vofupe1dqz9emx13r"],
          },
        },
        userInfo: {
          id: "johndoe",
          password: "123456",
          name: "John Doe",
          avatarURL: `${process.env.PUBLIC_URL}/imgs/avt_2.jpeg`,
          answers: {
            xj352vofupe1dqz9emx13r: "optionOne",
            vthrdm985a262al8qx3do: "optionTwo",
            "6ni6ok3ym7mf1p33lnez": "optionOne",
          },
          questions: ["6ni6ok3ym7mf1p33lnez", "xj352vofupe1dqz9emx13r"],
        },
        isAuth: true,
      },
    };

    jest
      .spyOn(redux, "useSelector")
      .mockImplementation((callback) => callback(mockStateRedux));
  });
  test("Render Detail Poll Page", () => {
    const { wrapper } = renderWithAllProvider(
      <Routes>
        <Route path={"/questions/:question_id"} element={<DetailPollPage />} />
      </Routes>,
      {
        initialEntries: ["/questions/am8ehyc8byjqgar0jgpub9"],
      }
    );
    const expectElement = screen.getByText(/Would You Rather?/i);
    expect(expectElement).toBeInTheDocument();
    expect(wrapper).toMatchSnapshot();
  });
  test("Leave Detail Poll Page", () => {
    const { wrapper } = renderWithAllProvider(
      <Routes>
        <Route
          path={"/questions/:question_id"}
          element={
            <HomeLayout>
              <DetailPollPage />
            </HomeLayout>
          }
        />
        <Route path={"/"} element={<HomeLayout></HomeLayout>} />
      </Routes>,
      {
        initialEntries: ["/questions/am8ehyc8byjqgar0jgpub9"],
      }
    );
    const elementLeaderBoard = screen.getByText(/Home/i);
    act(() => {
      elementLeaderBoard.click();
    });
    expect(screen.queryByText(/Would You Rather?/i)).not.toBeInTheDocument();
    expect(wrapper).toMatchSnapshot();
  });
  test("Render Poll by sarahedo", () => {
    const { wrapper } = renderWithAllProvider(
      <Routes>
        <Route path={"/questions/:question_id"} element={<DetailPollPage />} />
      </Routes>,
      {
        initialEntries: ["/questions/am8ehyc8byjqgar0jgpub9"],
      }
    );
    const elementPollByUser = screen.getByText(/Poll by sarahedo/i);
    expect(elementPollByUser).toBeInTheDocument();
    expect(wrapper).toMatchSnapshot();
  });
  test("Render option One", () => {
    const { wrapper } = renderWithAllProvider(
      <Routes>
        <Route path={"/questions/:question_id"} element={<DetailPollPage />} />
      </Routes>,
      {
        initialEntries: ["/questions/am8ehyc8byjqgar0jgpub9"],
      }
    );
    const elementOptionOne = screen.getByText(/be telekinetic/i);
    expect(elementOptionOne).toBeInTheDocument();
    expect(wrapper).toMatchSnapshot();
  });
  test("Render option Two", () => {
    const { wrapper } = renderWithAllProvider(
      <Routes>
        <Route path={"/questions/:question_id"} element={<DetailPollPage />} />
      </Routes>,
      {
        initialEntries: ["/questions/am8ehyc8byjqgar0jgpub9"],
      }
    );
    const elementOptionTwo = screen.getByText(/be telepathic/i);
    expect(elementOptionTwo).toBeInTheDocument();
    expect(wrapper).toMatchSnapshot();
  });
  test("Render logout btn", () => {
    const { wrapper } = renderWithAllProvider(
      <Routes>
        <Route
          path={"/questions/:question_id"}
          element={
            <HomeLayout>
              <DetailPollPage />
            </HomeLayout>
          }
        />
      </Routes>,
      {
        initialEntries: ["/questions/am8ehyc8byjqgar0jgpub9"],
      }
    );
    const elementLogoutBtn = screen.getByText(/Logout/i);
    expect(elementLogoutBtn).toBeInTheDocument();
    expect(wrapper).toMatchSnapshot();
  });
});
