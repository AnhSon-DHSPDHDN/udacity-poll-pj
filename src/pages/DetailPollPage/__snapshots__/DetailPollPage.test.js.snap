// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Detail Poll Page Leave Detail Poll Page 1`] = `undefined`;

exports[`Detail Poll Page Render Detail Poll Page 1`] = `undefined`;

exports[`Detail Poll Page Render Poll by sarahedo 1`] = `undefined`;

exports[`Detail Poll Page Render logout btn 1`] = `undefined`;

exports[`Detail Poll Page Render option One 1`] = `undefined`;

exports[`Detail Poll Page Render option Two 1`] = `undefined`;
